### DISCUSSION TITLE
Clarification on Bypass Behavior in find_victim Function

### DISCUSSION THREAD

**Discussion by yhj27** [ChampSim/ChampSim#381] [Q&A] [answered]
Hello, I am currently researching cache replacement policies using ChampSim. I have a specific question about the bypass behavior in ChampSim. In the **find_victim** function, if LLC_WAYS (e.g., 16 if the cache is 16-way associative) is returned, a bypass operation is triggered.

Could you please clarify what happens to the data that is bypassed in this case? Does it get stored in L2 or L1, or does it go directly back to the Core? Also, does this behavior differ depending on the access types (Load, RFO, Prefetch)? I would greatly appreciate it if you could provide explanations for each case.

Thank you!

**Comment by ngober** [comment, marked as answer]
We should answer this question in two parts: First, the case where the upper level did not request a return, and second where it did.

### Upper Level Did Not Request Return

This happens in two cases: for writes/writebacks (as a result of upper level evictions of dirty blocks, or store requests in the highest level), and non-filling prefetches (with `prefetch_line(*, false, *)`).

Bypassing a writeback violates correctness, and is [protected by an assertion.](https://github.com/ChampSim/ChampSim/blob/ab68a5e6cf626b014422f6c9ae09a2fbcd549880/src/cache.cc#L147) If this occurs, the simulation will crash.

Bypassing a non-filling prefetch may be an error, but the simulator does not prevent it. This is a situation where the user-provided prefetcher and user-provided replacement disagree on what action should be performed on the block. You should handle these cases on your own. The simulator will not help you.

### Upper Level Requested Return

In these cases, all of the packets are load-like requests. The packet will be returned to the upper level, but will not be filled to this level. When the packet is returned to the upper level, the replacement policy there will decide if the packet should be filled. There is no communication of the bypass to the upper level.
